test/unittest/agent/AgentUnitTests.cpp

```diff
@@ -0,0 +1,962 @@
+// Copyright 2019 Proyectos y Sistemas de Mantenimiento SL (eProsima).
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <uxr/agent/Agent.hpp>
+
+#include <gtest/gtest.h>
+
+namespace eprosima {
+namespace uxr {
+namespace testing {
+
+using namespace eprosima::uxr;
+
+class AgentUnitTests : public ::testing::Test
+{
+protected:
+    AgentUnitTests() = default;
+
+    ~AgentUnitTests()
+    {
+        eprosima::uxr::Agent::reset();
+    }
+
+    uint32_t client_key_ = 0xAABBCCDD;
+};
+
+TEST_F(AgentUnitTests, CreateClient)
+{
+    Agent::ResultStatus result;
+    EXPECT_TRUE(Agent::create_client(client_key_, 0x01, 512, result));
+    EXPECT_TRUE(Agent::create_client(client_key_, 0x01, 512, result));
+}
+
+TEST_F(AgentUnitTests, DeleteClient)
+{
+    Agent::ResultStatus result;
+    Agent::create_client(client_key_, 0x01, 512, result);
+
+    EXPECT_TRUE(Agent::delete_client(client_key_, result));
+    EXPECT_FALSE(Agent::delete_client(client_key_, result));
+    EXPECT_EQ(result, Agent::ResultStatus::UNKNOWN_REFERENCE_ERROR);
+}
+
+TEST_F(AgentUnitTests, CreateParticipantByRef)
+{
+    Agent::ResultStatus result;
+    Agent::create_client(client_key_, 0x01, 512, result);
+
+    const char* ref_one = "default_xrce_participant";
+    const char* ref_two = "default_xrce_participant_two";
+
+    uint16_t participant_id = Agent::get_object_id(0x00, Agent::PARTICIPANT_OBJK);
+    int16_t domain_id = 0x00;
+    uint8_t flag = 0x00;
+
+    /*
+     * Create Participant.
+     */
+    EXPECT_TRUE(Agent::create_participant_by_ref(client_key_, participant_id, domain_id, ref_one, flag, result));
+
+    /*
+     * Create Participant over an existing with 0x00 flag.
+     */
+    EXPECT_FALSE(Agent::create_participant_by_ref(client_key_, participant_id, domain_id, ref_one, flag, result));
+    EXPECT_EQ(result, Agent::ResultStatus::ALREADY_EXISTS_ERROR);
+    EXPECT_FALSE(Agent::create_participant_by_xml(client_key_, participant_id, domain_id, ref_two, flag, result));
+    EXPECT_EQ(result, Agent::ResultStatus::ALREADY_EXISTS_ERROR);
```

Comment:
Are you sure this is the expected error? The entity is created by XML with a reference as XML parameter. Do you want to put `_ref`?

Reply:
well seeing!